### PR TITLE
server : (refactor) no more json in server_task input

### PR DESCRIPTION
**Pull request by ngxson**
Continue https://github.com/ggerganov/llama.cpp/pull/10643

`server_task_result` is already broken into multiple derived classes (polymorphism). This helps reduce code complexity because each of the result type is different from another.

However, the `server_task` can't be benefit from the same approach, because most requests share the same parameters with other.

The solution introduced by this PR is to just put everything into `server_task`. Also the JSON parsing is now done at HTTP thread. Up on receiving a request, HTTP thread parse JSON into one or more `server_task` and push them to `server_queue`

---

Example of `/slots` response:

```
[
  {
    "id": 0,
    "id_task": -1,
    "n_ctx": 1024,
    "speculative": false,
    "is_processing": false,
    "params": {
      "n_predict": -1,
      "seed": 4294967295,
      "temperature": 0.800000011920929,
      "dynatemp_range": 0.0,
      "dynatemp_exponent": 1.0,
      "top_k": 40,
      "top_p": 0.949999988079071,
      "min_p": 0.05000000074505806,
      "xtc_probability": 0.0,
      "xtc_threshold": 0.10000000149011612,
      "typical_p": 1.0,
      "repeat_last_n": 64,
      "repeat_penalty": 1.0,
      "presence_penalty": 0.0,
      "frequency_penalty": 0.0,
      "dry_multiplier": 0.0,
      "dry_base": 1.75,
      "dry_allowed_length": 2,
      "dry_penalty_last_n": -1,
      "dry_sequence_breakers": [
        "\n",
        ":",
        "\"",
        "*"
      ],
      "mirostat": 0,
      "mirostat_tau": 5.0,
      "mirostat_eta": 0.10000000149011612,
      "penalize_nl": false,
      "stop": [],
      "max_tokens": -1,
      "n_keep": 0,
      "n_discard": 0,
      "ignore_eos": false,
      "stream": true,
      "n_probs": 0,
      "min_keep": 0,
      "grammar": "",
      "samplers": [
        "dry",
        "top_k",
        "typ_p",
        "top_p",
        "min_p",
        "xtc",
        "temperature"
      ],
      "speculative.n_max": 16,
      "speculative.n_min": 5,
      "speculative.p_min": 0.8999999761581421,
      "timings_per_token": false
    },
    "prompt": "",
    "next_token": {
      "has_next_token": true,
      "has_new_line": false,
      "n_remain": -1,
      "n_decoded": 0,
      "stopping_word": ""
    }
  }
]
```

Example of `/props` response:

```
{
  "default_generation_settings": {
    "id": 0,
    "id_task": -1,
    "n_ctx": 1024,
    "speculative": false,
    "is_processing": false,
    "params": {
      "n_predict": -1,
      "seed": 4294967295,
      "temperature": 0.800000011920929,
      "dynatemp_range": 0.0,
      "dynatemp_exponent": 1.0,
      "top_k": 40,
      "top_p": 0.949999988079071,
      "min_p": 0.05000000074505806,
      "xtc_probability": 0.0,
      "xtc_threshold": 0.10000000149011612,
      "typical_p": 1.0,
      "repeat_last_n": 64,
      "repeat_penalty": 1.0,
      "presence_penalty": 0.0,
      "frequency_penalty": 0.0,
      "dry_multiplier": 0.0,
      "dry_base": 1.75,
      "dry_allowed_length": 2,
      "dry_penalty_last_n": -1,
      "dry_sequence_breakers": [
        "\n",
        ":",
        "\"",
        "*"
      ],
      "mirostat": 0,
      "mirostat_tau": 5.0,
      "mirostat_eta": 0.10000000149011612,
      "penalize_nl": false,
      "stop": [],
      "max_tokens": -1,
      "n_keep": 0,
      "n_discard": 0,
      "ignore_eos": false,
      "stream": true,
      "n_probs": 0,
      "min_keep": 0,
      "grammar": "",
      "samplers": [
        "dry",
        "top_k",
        "typ_p",
        "top_p",
        "min_p",
        "xtc",
        "temperature"
      ],
      "speculative.n_max": 16,
      "speculative.n_min": 5,
      "speculative.p_min": 0.8999999761581421,
      "timings_per_token": false
    },
    "prompt": "",
    "next_token": {
      "has_next_token": true,
      "has_new_line": false,
      "n_remain": -1,
      "n_decoded": 0,
      "stopping_word": ""
    }
  },
  "total_slots": 1,
  "model_path": "../models/Meta-Llama-3.1-8B-Instruct-Q4_K_M.gguf",
  "chat_template": "..."
}
```

### CODE REVIEW

File: examples/server/server.cpp [hunk 1]
```diff
@@ -186,6 +165,212 @@ struct slot_params {
...
+        std::string filepath;
+    };
+    slot_action slot_action;
+
+    // used by SERVER_TASK_TYPE_METRICS
+    bool metrics_reset_bucket = false;
+
+    server_task(
+        server_task_type type,
+        server_task_inf_type inf_type = SERVER_TASK_INF_TYPE_COMPLETION) : type(type), inf_type(inf_type) {}
```
Comment (ggerganov) [inline examples/server/server.cpp:196]: It's better to be explicit here about the inference type:

```suggestion
    server_task(
        server_task_type type,
        server_task_inf_type inf_type) : type(type), inf_type(inf_type) {}
```

File: examples/server/server.cpp [hunk 2]
```diff
@@ -118,6 +96,7 @@ struct slot_params {
 
     std::vector<std::string> antiprompt;
     bool timings_per_token = false;
+    bool ignore_eos = false;
```
Comment (ggerganov) [inline examples/server/server.cpp:99]: With new models the `ignore_eos` functionality is losing relevance. There are now many different "end-of-generation" tokens and it's not just a single EOS token anymore. We should remove this logic and only support logit biases, which is more general. Just a note, no need to do it in this PR.

File: examples/server/server.cpp [hunk 3]
```diff
@@ -3416,14 +3358,41 @@ int main(int argc, char ** argv) {
...
-        data["completion_id"] = gen_chatcmplid();
-        std::vector<server_task> tasks = ctx_server.create_tasks_inference(data, inf_type);
+        auto completion_id = gen_chatcmplid();
+        std::vector<server_task> tasks;
+
+        try {
+            std::vector<llama_tokens> tokenized_prompts = tokenize_input_prompts(ctx_server.ctx, data.at("prompt"), true, true);
+            tasks.reserve(tokenized_prompts.size());
+            for (size_t i = 0; i < tokenized_prompts.size(); i++) {
+                server_task task = server_task(SERVER_TASK_TYPE_INFERENCE, inf_type);
```
Comment (ggerganov) [inline examples/server/server.cpp:3375]:
```suggestion
                server_task task(SERVER_TASK_TYPE_INFERENCE, inf_type);
```

File: examples/server/server.cpp [hunk 4]
```diff
@@ -3638,7 +3616,16 @@ int main(int argc, char ** argv) {
         json responses = json::array();
         bool error = false;
         {
-            std::vector<server_task> tasks = ctx_server.create_tasks_inference({{"prompt", prompt}}, SERVER_TASK_INF_TYPE_EMBEDDING);
+            std::vector<server_task> tasks;
+            std::vector<llama_tokens> tokenized_prompts = tokenize_input_prompts(ctx_server.ctx, prompt, /* add_special */ false, true);
+            for (size_t i = 0; i < tokenized_prompts.size(); i++) {
+                server_task task = server_task(SERVER_TASK_TYPE_INFERENCE, SERVER_TASK_INF_TYPE_EMBEDDING);
```
Comment (ggerganov) [inline examples/server/server.cpp:3622]:
```suggestion
                server_task task(SERVER_TASK_TYPE_INFERENCE, SERVER_TASK_INF_TYPE_EMBEDDING);
```

File: examples/server/server.cpp [hunk 5]
```diff
@@ -3704,20 +3691,23 @@ int main(int argc, char ** argv) {
...
         // create and queue the task
         json responses = json::array();
         bool error = false;
         {
-            std::vector<server_task> tasks = ctx_server.create_tasks_inference({{"prompt", prompt}}, SERVER_TASK_INF_TYPE_RERANK);
+            std::vector<server_task> tasks;
+            std::vector<llama_tokens> tokenized_docs = tokenize_input_prompts(ctx_server.ctx, documents, /* add_special */ false, true);
+            tasks.reserve(tokenized_docs.size());
+            for (size_t i = 0; i < tokenized_docs.size(); i++) {
+                server_task task = server_task(SERVER_TASK_TYPE_INFERENCE, SERVER_TASK_INF_TYPE_RERANK);
```
Comment (ggerganov) [inline examples/server/server.cpp:3704]:
```suggestion
                server_task task(SERVER_TASK_TYPE_INFERENCE, SERVER_TASK_INF_TYPE_RERANK);
```

**Review by ggerganov — APPROVED** [review]

File: examples/server/server.cpp (hunk 1 above)
Comment (ngxson) [reply examples/server/server.cpp:196]: the inf_type is only required with `SERVER_TASK_TYPE_INFERENCE`, so I think we should provide a default value for it when we don't want to use, right? For example, `server_task(SERVER_TASK_TYPE_METRICS)`

Another way is to remove `server_task_inf_type inf_type` from the constructor and set it later on.

Comment (ggerganov) [reply examples/server/server.cpp:196]: The default value should be something like `SEREVR_TASK_INF_TYPE_NONE`.

Though I think it's better to combine the `server_task_type` and `server_task_inf_type` enums in a single enum:

```c++
enum server_task_type {
    SERVER_TASK_TYPE_COMPLETION,
    SERVER_TASK_TYPE_EMBEDDING,
    SERVER_TASK_TYPE_RERANK,
    SERVER_TASK_TYPE_INFILL,
    SERVER_TASK_TYPE_CANCEL,
    SERVER_TASK_TYPE_NEXT_RESPONSE,
    SERVER_TASK_TYPE_METRICS,
    SERVER_TASK_TYPE_SLOT_SAVE,
    SERVER_TASK_TYPE_SLOT_RESTORE,
    SERVER_TASK_TYPE_SLOT_ERASE,
    SERVER_TASK_TYPE_SET_LORA,
};
```

Comment (ngxson) [reply examples/server/server.cpp]: I ended up combining them in https://github.com/ggerganov/llama.cpp/pull/10691/commits/090a113417117b90c7c8a946316caa6d9cba7544

**Review by ggerganov — APPROVED** [review]

**Comment by ggerganov** [comment]
This change breaks the infill endpoint - it produces mostly garbage.

**Comment by ngxson** [comment]
Hmm ok could be due to the infill "template" is not being applied correctly. I'll add a test with qwen model (run locally, not on CI)

**Comment by ngxson** [comment]
I'm on it, will make a PR

**Pull request merged**